sb_vision/find_3D_coords.py

```diff
@@ -0,0 +1,139 @@
+        data = cast(List[List[np.float64]], np.reshape(
+            [float(v) for v in values],
+            (rows, cols),
+        ).tolist())
+        return data
+    else:
+        raise ValueError('Unexpected type_id of tag ({})'.format(type_id))
+
+
+def get_calibration(file_name: Path) -> Tuple[List[List[float]], List[List[float]]]:
+    """
+    Parse a calibration xml generated by the camera calibration tool.
+
+    (see https://docs.opencv.org/3.4.0/d7/d21/tutorial_interactive_calibration.html)
+    :param file_name: name of xml file to parse
+    :return: calibrations from the file in a dictionary
+    """
+    with file_name.open() as file:
+        tree = etree.parse(file)
+        camera_matrix = convert(tree.find('cameraMatrix'))
+        dist_coeffs = convert(tree.find('dist_coeffs'))
+
+    return camera_matrix, dist_coeffs
+
+
+@functools.lru_cache()
+def load_camera_calibrations(camera_model: str) -> Tuple[List[List[float]],
+                                                         List[List[float]]]:
+    """
+    Load camera calibrations from a file.
+
+    :param camera_model: file to load
+    :return: camera calibrations
+    """
+    builtin_models_dir = Path(__file__).parent
+    model_file = builtin_models_dir / '{}_calibration.xml'.format(
+        camera_model,
+    )
+
+    camera_matrix, distance_coefficients = get_calibration(model_file)
+    return camera_matrix, distance_coefficients
+
+
+def calculate_transforms(
+    marker_size: Tuple[float, float],
+    pixel_corners: List[PixelCoordinate],
+    camera_matrix: List[List[float]],
+    distance_coefficients: List[List[float]],
+) -> Tuple[Cartesian, Tuple[float, float, float]]:
+    """
+    Calculate the position of a marker.
+
+    Given the pixel co-ordinates of the corners and the calibrations
+    of the camera.
+
+    :param marker_size: size of the marker
+    :param pixel_corners: pixel co-ordinates of the corners of the marker
+        (clockwise around the marker from the top-left corner)
+    :param camera_matrix: calibration matrix for the camera
+    :param distance_coefficients: distance calibration for the camera
+    :return: translation and orientation of the marker
+    """
+    w, h = marker_size
+    width_from_centre = w / 2
+    height_from_centre = h / 2
+
+    # create the rectangle representing the marker in 3D
+    object_points = np.array([
+        [width_from_centre, height_from_centre, 0],
+        [width_from_centre, -height_from_centre, 0],
+        [-width_from_centre, -height_from_centre, 0],
+        [-width_from_centre, height_from_centre, 0],
+    ])
+
+    return_value, orientation_vector, translation_vector = cv2.solvePnP(
+        object_points,
+        np.array(pixel_corners),
+        np.array(camera_matrix),
+        np.array(distance_coefficients),
+    )
+    if not return_value:
+        raise ValueError("cv2.solvePnP returned false")
+
+    translation_vector = Cartesian(*(v[0] for v in translation_vector))
```

Comment:
I don't think it's a good idea to put values into this typed named tuple which don't match the contract of the type.

Reply:
I don't really understand what you mean here, the type is a triplet of AnyFloat, and I'm passing in a triplet of floats, in what way does it not match the contract?

Reply:
The contract is not just that it's a type -- we also expect that positive y means up. However the conversion which follows indicates that that wasn't actually the case here. I realise that that's not actually documented on the type itself, though the README does specify this.

This actually goes away in my `solvePnP` wrapper (#36) as I've pushed this fix down into the C layer.